src/Microsoft.AspNet.TestHost/ClientHandler.cs

```diff
@@ -70,33 +70,36 @@ public ClientHandler([NotNull] Func<IFeatureCollection, Task> next, PathString p
                     try
                     {
                         await _next(state.FeatureCollection);
-                        state.CompleteResponse();
+                        state.PipelineComplete();
                     }
                     catch (Exception ex)
                     {
-                        state.Abort(ex);
+                        state.PipelineFailed(ex);
                     }
                     finally
                     {
                         registration.Dispose();
-                        state.Dispose();
                     }
                 });
 
             return await state.ResponseTask;
         }
 
-        private class RequestState : IDisposable
+        private class RequestState
         {
             private readonly HttpRequestMessage _request;
             private TaskCompletionSource<HttpResponseMessage> _responseTcs;
             private ResponseStream _responseStream;
             private ResponseFeature _responseFeature;
+            private CancellationTokenSource _requestAbortedSource;
+            private bool _pipelineFinished;
 
             internal RequestState(HttpRequestMessage request, PathString pathBase, CancellationToken cancellationToken)
```

Comment:
There is a registration in SendAsync which calls AbortRequest, which then Cancels _requestAbortedSource. AbortRequest also calls Complete on the _responseStream.